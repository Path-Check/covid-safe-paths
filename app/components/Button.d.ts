import { Button as button } from './Button';

export const Button: (props: Record<string, unknown>) => JSX.Element = button;
