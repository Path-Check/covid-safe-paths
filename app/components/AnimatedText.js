import React, { useEffect, useState } from 'react';

/**
 * Takes an input label and animates a 3 dot suffix, updating each second.
 *
 * Example:
 *
 * Loading
 *
 * Loading.
 *
 * Loading..
 *
 * Loading...
 *
 * Loading
 *
 * Loading.
 * @param {string} label
 * @param {number} interval
 */
export const AnimatedText = ({ value, interval }) => {
  const [text, setText] = useState(value);

  useEffect(() => {
    const labelInterval = setInterval(() => {
      if (text.slice(-3) === '...') {
        setText(value);
      } else {
        setText(text + '.');
      }
    }, interval);

    return () => clearInterval(labelInterval);
  }, [interval, setText, text, value]);

  return <>{text}</>;
};
