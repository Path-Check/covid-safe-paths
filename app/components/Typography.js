import * as React from 'react';
import styled from '@emotion/native';

export const Type = {
  Headline1: 'headline1',
  Headline2: 'headline2',
  Body1: 'body1',
  Body2: 'body2',
  Body3: 'body3',
};

/**
 * A correctly sized, spaced and colored text element.
 *
 * Use within a `<ThemeProvider>`
 *
 * @param {{
 *   use?: string,
 *   secondary?: boolean,
 *   monospace?: boolean,
 * }} param0
 */
export const Typography = ({
  use = Type.Body1,
  secondary,
  monospace,
  children,
  ...otherProps
}) => {
  return (
    <ThemedText
      use={use}
      secondary={secondary}
      monospace={monospace}
      {...otherProps}>
      {children}
    </ThemedText>
  );
};

const FONT_SIZE_MAP = {
  [Type.Headline1]: '52px',
  [Type.Headline2]: '26px',
  [Type.Body1]: '18px',
  [Type.Body2]: '16px',
  [Type.Body3]: '15px',
};

const getFontSize = ({ use = Type.Body1 }) => FONT_SIZE_MAP[use];

const LINE_HEIGHT_MAP = {
  [Type.Headline1]: '48px',
  [Type.Headline2]: '34px',
  [Type.Body1]: '24px',
  [Type.Body2]: '22px',
  [Type.Body3]: '24px',
};

const getLineHeight = ({ use = Type.Body1 }) => LINE_HEIGHT_MAP[use];

const getTextColor = ({ theme, secondary = false }) =>
  secondary ? theme.textSecondaryOnBackground : theme.textPrimaryOnBackground;

const getFontWeight = ({ use = Type.Body1 }) =>
  use.startsWith('headline') ? 'bold' : 'normal';

const ThemedText = styled.Text`
  font-family: ${({ monospace }) =>
    monospace ? 'IBM Plex Mono' : 'IBM Plex Sans'};
  font-size: ${getFontSize};
  font-weight: ${getFontWeight};
  line-height: ${getLineHeight};
  color: ${getTextColor};
`;
