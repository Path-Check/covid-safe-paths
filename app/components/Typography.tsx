import React from 'react';
import { Text, TextStyle } from 'react-native';

import { useLanguageDirection } from '../locales/languages';

import { Typography as TypographyStyles } from '../styles';

type TypographicUse = 'headline1' | 'headline2' | 'body1' | 'body2' | 'body3';

interface TypographyProps {
  use?: TypographicUse;
  style?: TextStyle;
  testID?: string;
  children: JSX.Element | string;
  onPress?: () => void;
}

export const Typography = ({
  use = 'body1',
  style,
  testID,
  children,
  onPress,
}: TypographyProps): JSX.Element => {
  const writingDirection = useLanguageDirection();

  const useToStyle = () => {
    switch (use) {
      case 'headline1': {
        return TypographyStyles.header1;
      }
      case 'headline2': {
        return TypographyStyles.header2;
      }
      case 'body1': {
        return TypographyStyles.mainContent;
      }
      case 'body2': {
        return TypographyStyles.secondaryContent;
      }
      case 'body3': {
        return TypographyStyles.tertirayContent;
      }
    }
  };

  const textStyle = useToStyle();

  return (
    <Text
      onPress={onPress}
      style={[textStyle, { writingDirection }, style]}
      testID={testID}>
      {children}
    </Text>
  );
};
