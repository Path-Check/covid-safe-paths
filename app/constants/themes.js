import React from 'react';
import styled from '@emotion/native';
import { ThemeProvider } from 'emotion-theming';

import Color from './colors';

/** Main theme, violet on pale violet bg. e.g. Settings */
export const main = {
  background: Color.VIOLET_ALPHA_06,
  textPrimaryOnBackground: Color.VIOLET,
  textSecondaryOnBackground: 'rgba(64, 81, 219, 0.6)',

  primary: Color.VIOLET_BUTTON,
  onPrimary: Color.WHITE,

  success: Color.SUCCESS,
  warning: Color.WARNING,
  borders: Color.GRAY_BACKGROUND,
};

/** Inverted theme, white on violet bg. E.g. Main screen */
export const inverted = {
  ...main,
  background: `linear-gradient(245.21deg, ${Color.VIOLET_BUTTON} 0%, ${Color.VIOLET_BUTTON_DARK} 100%);`,
  textPrimaryOnBackground: Color.WHITE,
  textSecondaryOnBackground: 'rgba(255, 255, 255, 0.6)',

  primary: Color.WHITE,
  onPrimary: Color.VIOLET,
};

/** Mono, black on white. e.g. Exposure history screen */
export const mono = {
  ...main,
  background: Color.WHITE,
  textPrimaryOnBackground: Color.MONO_DARK,
  textSecondaryOnBackground: Color.MONO_SECONDARY,

  primary: Color.VIOLET_BUTTON,
  onPrimary: Color.WHITE,
};

/** Dark, white on gray. E.g. Possible exposure mode on main screen */
export const dark = {
  ...main,
  background: Color.DARK_GRAY,
  textPrimaryOnBackground: Color.WHITE,
  textSecondaryOnBackground: Color.LIGHT_GRAY,

  primary: Color.WHITE,
  onPrimary: Color.DARK_GRAY,
};

const THEME_MAP = {
  main,
  inverted,
  mono,
  dark,
};

export const ThemedBackground = styled.View`
  background-color: ${({ theme }) => theme.background};
`;

/** Return theme config for a known theme name */
export const getTheme = t => {
  const themeColors = THEME_MAP[t];
  if (!themeColors) {
    throw new Error(`Unknown theme ${t}`);
  }
  return themeColors;
};

/**
 * Configure themeable components with primary, bg and secondary colors
 *
 * @param {{
 *   use?: string,
 *   setBackground?: boolean,
 * }} param0
 */
export const Theme = ({ use = 'main', setBackground = false, children }) => {
  return (
    <ThemeProvider theme={getTheme(use)}>
      {setBackground ? (
        <ThemedBackground>{children}</ThemedBackground>
      ) : (
        children
      )}
    </ThemeProvider>
  );
};
