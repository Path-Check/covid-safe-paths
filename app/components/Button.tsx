import React from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  ViewStyle,
  TouchableOpacity,
} from 'react-native';

import { Typography } from './Typography';

import { Buttons, Typography as TypographyStyles } from '../styles';

interface ButtonProps {
  label: string;
  onPress: () => void;
  loading?: boolean;
  disabled?: boolean;
  style?: ViewStyle;
}

export const Button = ({
  label,
  onPress,
  disabled,
  loading,
  style,
}: ButtonProps): JSX.Element => {
  const textStyle =
    disabled || loading ? styles.textDisabled : styles.textEnabled;

  return (
    <TouchableOpacity
      onPress={onPress}
      accessible
      accessibilityLabel={label}
      accessibilityRole='button'
      disabled={disabled || loading}
      style={[styles.button, style]}>
      {loading ? (
        <ActivityIndicator size={'large'} />
      ) : (
        <Typography style={textStyle}>{label}</Typography>
      )}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    ...Buttons.largeWhite,
  },
  textEnabled: {
    ...TypographyStyles.buttonTextDark,
  },
  textDisabled: {
    ...TypographyStyles.buttonTextDark,
  },
});
