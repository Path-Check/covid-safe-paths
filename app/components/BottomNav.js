import React, { useState, useEffect } from 'react';
import * as NavIcons from '../assets/svgs/BottomNav';
import { Theme } from '../constants/themes';
import Colors from '../constants/colors';
import { View, StyleSheet, TouchableOpacity } from 'react-native';
import { SvgXml } from 'react-native-svg';
import { Typography } from '.';
import { useNavigation, useRoute } from '@react-navigation/native';
import { isGPS } from '../COVIDSafePathsConfig';
import { useTranslation } from 'react-i18next';
import LocationService from '../services/LocationService';

const NavIconButton = ({ icons, label, screens, hasNotifications }) => {
  const [hasExposure, setHasExposure] = useState(false);

  const navigation = useNavigation();
  const route = useRoute();

  const isActive =
    screens.primary === route.name || screens.children.includes(route.name);

  const icon = icons[isActive ? 'active' : 'inactive'];

  useEffect(() => {
    const checkForExposure = async () => {
      setHasExposure(await LocationService.getHasPotentialExposure());
    };

    // The History nav icon is the only element that needs to check for exposure
    // and create a notification element if the user has been exposed.
    if (screens.primary === 'ExposureHistoryScreen') {
      checkForExposure();
    }
  });

  return (
    <Theme use={isActive ? 'violet' : 'default'}>
      <TouchableOpacity
        onPress={() => {
          navigation.navigate(screens.primary);
        }}
        style={
          hasNotifications
            ? [styles.navIconRoot, { paddingBottom: 7 }]
            : styles.navIconRoot
        }>
        {hasExposure && (
          <SvgXml
            xml={NavIcons.NotificationIcon}
            width={8}
            height={8}
            style={styles.notificationIcon}
          />
        )}
        <SvgXml xml={icon} width={24} height={24} />
        <Typography use='Body1'>{label}</Typography>
      </TouchableOpacity>
    </Theme>
  );
};

export const BottomNav = () => {
  const { t } = useTranslation();

  const HistoryNav = {
    icons: {
      active: NavIcons.HistoryActive,
      inactive: NavIcons.HistoryInactive,
    },
    label: t('navigation.history'),
    screens: { primary: 'ExposureHistoryScreen', children: [] },
  };

  const HomeNav = {
    icons: { active: NavIcons.HomeActive, inactive: NavIcons.HomeInactive },
    label: t('navigation.home'),
    screens: { primary: 'Main', children: [] },
  };

  const LocationsNav = {
    icons: {
      active: NavIcons.LocationsActive,
      inactive: NavIcons.LocationsInactive,
    },
    label: t('navigation.locations'),
    screens: { primary: 'ExportScreen', children: [] }, // TODO: Determine if this is correct
  };

  const MoreNav = {
    icons: { active: NavIcons.MoreActive, inactive: NavIcons.MoreInactive },
    label: t('navigation.more'),
    screens: {
      primary: 'SettingsScreen',
      children: ['AboutScreen', 'LicensesScreen', 'FeatureFlagsScreen'],
    },
  };

  const PartnersNav = {
    icons: {
      active: NavIcons.PartnersActive,
      inactive: NavIcons.PartnersInactive,
    },
    label: t('navigation.partners'),
    screens: { primary: 'ChooseProviderScreen', children: [] },
  };

  const gpsNavIcons = [HomeNav, HistoryNav, LocationsNav, PartnersNav, MoreNav];

  // Excludes `LocationsNav`
  const bteNavIcons = [HomeNav, HistoryNav, PartnersNav, MoreNav];

  const navIconButtons = isGPS ? gpsNavIcons : bteNavIcons;

  return (
    <View style={styles.navRoot}>
      {navIconButtons.map((navIconButton) => (
        <NavIconButton key={navIconButton.label} {...navIconButton} />
      ))}
    </View>
  );
};

const styles = StyleSheet.create({
  navRoot: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    height: 60,
    backgroundColor: Colors.VIOLET_TEXT_DARK,
  },
  navIconRoot: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  notificationIcon: {
    top: 3,
    left: 18,
  },
});
