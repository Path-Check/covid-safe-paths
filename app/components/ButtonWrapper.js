import * as React from 'react';
import { StyleSheet, Dimensions, View } from 'react-native';
import Button from './Button';
import PropTypes from 'prop-types';
import fontFamily from '../constants/fonts';

const width = Dimensions.get('window').width;

class ButtonWrapper extends React.Component {
  render() {
    return (
      <View style={styles.buttonContainer}>
        <Button titleStyle={styles.primaryButtonText} {...this.props} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  primaryButtonText: {
    fontFamily: fontFamily.primarySemiBold,
    fontSize: 16,
    lineHeight: 19,
    letterSpacing: 0,
    textAlign: 'center',
    color: '#ffffff',
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    width: width * 0.8,
    alignSelf: 'center',
  },
});

ButtonWrapper.propTypes = {
  title: PropTypes.string.isRequired,
  onPress: PropTypes.func.isRequired,
  buttonColor: PropTypes.string,
  bgColor: PropTypes.string.isRequired,
  toBgColor: PropTypes.string,
  titleStyle: PropTypes.object,
  buttonStyle: PropTypes.object,
  borderColor: PropTypes.string,
};

export default ButtonWrapper;
