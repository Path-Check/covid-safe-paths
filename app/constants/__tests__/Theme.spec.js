import React from 'react';
import { render } from '@testing-library/react-native';

import { Theme } from '../themes';
import { Typography } from '../../components/Typography';

it('includes extra background View if setBackground=true', () => {
  const { asJSON } = render(
    <Theme use='dark' setBackground>
      <Typography use='headline1'>Text</Typography>
    </Theme>,
  );

  expect(asJSON()).toMatchSnapshot();
});

it('changes text color based on theme', () => {
  const { asJSON } = render(
    <>
      <Theme use='dark'>
        <Typography use='body1'>Text</Typography>
      </Theme>
      <Theme use='main'>
        <Typography use='body1'>Text</Typography>
      </Theme>
    </>,
  );

  expect(asJSON()).toMatchSnapshot();
});
