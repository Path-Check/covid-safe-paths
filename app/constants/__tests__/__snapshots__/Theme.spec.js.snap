// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes text color based on theme 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#2e2e2e",
          "fontFamily": "IBMPlexSans",
          "fontSize": 17,
          "lineHeight": 28,
        },
        Object {
          "writingDirection": "ltr",
        },
        undefined,
      ]
    }
  >
    Text
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "#2e2e2e",
          "fontFamily": "IBMPlexSans",
          "fontSize": 17,
          "lineHeight": 28,
        },
        Object {
          "writingDirection": "ltr",
        },
        undefined,
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`includes extra background View if setBackground=true 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#4e4e4e",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#1f2c9b",
            "fontFamily": "IBMPlexSans-Bold",
            "fontSize": 52,
            "fontWeight": "500",
            "lineHeight": 52,
          },
          Object {
            "writingDirection": "ltr",
          },
          undefined,
        ]
      }
    >
      Text
    </Text>
  </View>
</View>
`;
